2018/10/16/src/main/java/me/lemire/microbenchmarks/algorithms/Utf8Validate.java

```diff
@@ -194,7 +196,7 @@ public static boolean isUTF8(byte[] b) {
     int length = b.length;
     int s = 0;
     for (int i = 0; i < length; i++) {
-      s = utf8d[256 + s + (utf8d[b[i] & 0xFF] & 0xFF) ];
+      s = utf8d_transition[(s + (utf8d_toclass[b[i] & 0xFF])) & 0xFF];
```

Comment:
This is the change that actually speeds things up: by moving the `0xFF` outside the other calculations (doing it at the end), the compiler knows the index value is in 0-255 and can eliminate the bounds check.

There is still an unnecessary `movzx` (move with zero extension) in the core loop, associated with the fact that `s` is an `int` but the table values are bytes - but I wasn't able to eliminate it (e.g., I tried making `s` a byte, but it just moved the problem elsewhere). The good news is that `movzx` is a "zero latency" instruction on most modern Intel (not on AMD according to Agner) chip, so you can still get a total 5 + 1 cycle latency even with that annoying instruction.